### ISSUE TITLE
ORKChoiceAnswerFormatHelper code review

### ISSUE THREAD

**Issue by jwe-apple** [ResearchKit/ResearchKit#270]
We need to review some of the code in ORKChoiceAnswerFormatHelper. It's set up to handle both answers that are boolean, and arrays of values for a text choice. However it needs to handle the case where the answer is an array that does not contain a valid choice. Just because there's no valid choice should not imply that we fall back to using integers.

**Comment by brenbob** [comment]
I'm seeing a (perhaps) related problem with multiple-choice ORKTextChoiceQuestion. For example:  
- I have a multiple-choice question with 6 choices,
- Each choice has a value of '1' if selected,

Per RK docs, choiceAnswers should be:
"An array of selected values, from the value property of an ORKAnswerOption object. In the case of a single choice, the array has exactly one entry."

So I expect only the second choice is selected with the below array:

questionDefault.choiceAnswers = [0,1]

But in fact, only the first choice is selected. 

Perhaps RK expects 1-based index values for selected items instead? But this test shows the code expects a zero-based array:

questionDefault.choiceAnswers = [2,4]

since the 3rd and 5th items are selected. Still, there's a problem with handling for the 1st and second choices, since this scenario:

questionDefault.choiceAnswers = [1,2]

results in the 1st and 3rd items being selected.

So I see two problems:

1 - mismatch between RK documentation and expected structure for the choiceAnswers array
2 - code can't correctly distinguish between 1st and 2nd choices

**Comment by jwe-apple** [comment]
@brenbob I see how this arises. For now, most of the problems should go away if you use string-based values rather than integers.

**Comment by brenbob** [comment]
Not sure if that works. Using text values as in these examples:

questionDefault.choiceAnswers = ["1"]
questionDefault.choiceAnswers = ["1","1"]

results in an error:

*** Assertion failure in -[ORKChoiceAnswerFormatHelper selectedIndexesForAnswer:]

**Comment by jwe-apple** [comment]
And "1" is a valid TextChoice value for that question?

**Comment by brenbob** [comment]
Yes. Each choice has a possible value represented in JSON like this: "value":"1" and I generate the formItem like this:

        var choices = [ORKTextChoice]()
        for option in options {
            let choiceText = option["text"] as! String
            let choiceValue = option["value"] as! String
            choices.append(ORKTextChoice(text: choiceText, value: choiceValue))
        }

**Comment by brenbob** [comment]
So it seems RK assumes each text choice will have a unique value. That's not the case in my survey, which helps explain the source of these errors.

**Comment by jwe-apple** [comment]
Yes, they need to be unique. Otherwise you wouldn't be able to distinguish the answers in the result. If we're not catching that they're not unique, we need to add a validator.

**Comment by brucehappy** [comment]
@jwe-apple I'll review ``ORKChoiceAnswerFormatHelper`` and check into the need for the validator
